chain/sync_manager.go

```diff
@@ -0,0 +1,359 @@
+	var bestBuck *syncTargetBucket
+	var bestTs *types.TipSet
+	for _, b := range sbs.buckets {
+		hts := b.heaviestTipSet()
+		if bestBuck == nil || bestTs.ParentWeight().LessThan(hts.ParentWeight()) {
+			bestBuck = b
+			bestTs = hts
+		}
+	}
+
+	sbs.removeBucket(bestBuck)
+
+	return bestBuck
+}
+
+func (sbs *syncBucketSet) removeBucket(toremove *syncTargetBucket) {
+	nbuckets := make([]*syncTargetBucket, 0, len(sbs.buckets)-1)
+	for _, b := range sbs.buckets {
+		if b != toremove {
+			nbuckets = append(nbuckets, b)
+		}
+	}
+	sbs.buckets = nbuckets
+}
+
+func (sbs *syncBucketSet) PopRelated(ts *types.TipSet) *syncTargetBucket {
+	for _, b := range sbs.buckets {
+		if b.sameChainAs(ts) {
+			sbs.removeBucket(b)
+			return b
+		}
+	}
+	return nil
+}
+
+func (sbs *syncBucketSet) Heaviest() *types.TipSet {
+	// TODO: should also consider factoring in number of peers represented by each bucket here
+	var bestTs *types.TipSet
+	for _, b := range sbs.buckets {
+		bhts := b.heaviestTipSet()
+		if bestTs == nil || bhts.ParentWeight().GreaterThan(bestTs.ParentWeight()) {
+			bestTs = bhts
+		}
+	}
+	return bestTs
+}
+
+type syncTargetBucket struct {
+	tips  []*types.TipSet
+	count int
+}
+
+func newSyncTargetBucket(tipsets ...*types.TipSet) *syncTargetBucket {
+	var stb syncTargetBucket
+	for _, ts := range tipsets {
+		stb.add(ts)
+	}
+	return &stb
+}
+
+func (stb *syncTargetBucket) sameChainAs(ts *types.TipSet) bool {
+	for _, t := range stb.tips {
+		if ts.Equals(t) {
+			return true
+		}
+		if types.CidArrsEqual(ts.Cids(), t.Parents()) {
+			return true
+		}
+		if types.CidArrsEqual(ts.Parents(), t.Cids()) {
+			return true
+		}
+	}
+	return false
+}
+
+func (stb *syncTargetBucket) add(ts *types.TipSet) {
+	stb.count++
+
+	for _, t := range stb.tips {
+		if t.Equals(ts) {
+			return
+		}
+	}
+
+	stb.tips = append(stb.tips, ts)
+}
+
+func (stb *syncTargetBucket) heaviestTipSet() *types.TipSet {
+	if stb == nil {
+		return nil
+	}
+
+	var best *types.TipSet
+	for _, ts := range stb.tips {
+		if best == nil || ts.ParentWeight().GreaterThan(best.ParentWeight()) {
+			best = ts
+		}
+	}
+	return best
+}
+
+func (sm *SyncManager) selectSyncTarget() (*types.TipSet, error) {
+	var buckets syncBucketSet
+
+	var peerHeads []*types.TipSet
+	for _, ts := range sm.peerHeads {
+		peerHeads = append(peerHeads, ts)
+	}
+	sort.Slice(peerHeads, func(i, j int) bool {
+		return peerHeads[i].Height() < peerHeads[j].Height()
+	})
+
+	for _, ts := range peerHeads {
+		buckets.Insert(ts)
+	}
+
+	if len(buckets.buckets) > 1 {
+		log.Warning("caution, multiple distinct chains seen during head selections")
+		// TODO: we *could* refuse to sync here without user intervention.
+		// For now, just select the best cluster
+	}
+
+	return buckets.Heaviest(), nil
+}
+
+func (sm *SyncManager) syncScheduler() {
+	var syncQueue syncBucketSet
+	var activeSyncTips syncBucketSet
+
+	var nextSyncTarget *syncTargetBucket
+	var workerChan chan *types.TipSet
+
+	for {
+		select {
```

Comment:
I'd put those cases in separate functions

Reply:
yeah... i was thinking i might have to do that. Will do it now